crates/services/gas_price_service/src/v1/service.rs

```diff
@@ -0,0 +1,507 @@
+
+        // when
+        service.run(&mut watcher).await.unwrap();
+        l2_block_sender.send(l2_block).await.unwrap();
+        service.shutdown().await.unwrap();
+
+        // then
+        let actual_price = read_algo.next_gas_price();
+        assert_ne!(initial_price, actual_price);
+    }
+
+    #[tokio::test]
+    async fn run__updates_gas_price_with_da_block_cost_source() {
+        // given
+        let block_height = 1;
+        let l2_block = BlockInfo::Block {
+            height: block_height,
+            gas_used: 60,
+            block_gas_capacity: 100,
+            block_bytes: 100,
+            block_fees: 100,
+        };
+
+        let (l2_block_sender, l2_block_receiver) = mpsc::channel(1);
+        let l2_block_source = FakeL2BlockSource {
+            l2_block: l2_block_receiver,
+        };
+
+        let metadata_storage = FakeMetadata::empty();
+        let config = V1AlgorithmConfig {
+            new_exec_gas_price: 100,
+            min_exec_gas_price: 50,
+            exec_gas_price_change_percent: 20,
+            l2_block_fullness_threshold_percent: 20,
+            gas_price_factor: NonZeroU64::new(10).unwrap(),
+            min_da_gas_price: 100,
+            max_da_gas_price_change_percent: 50,
+            da_p_component: 4,
+            da_d_component: 2,
+            normal_range_size: 10,
+            capped_range_size: 100,
+            decrease_range_size: 4,
+            block_activity_threshold: 20,
+        };
+        let (algo_updater, shared_algo) =
+            initialize_algorithm(&config, block_height, &metadata_storage).unwrap();
+
+        let notifier = Arc::new(tokio::sync::Notify::new());
+        let da_source = DaSourceService::new(
+            DummyDaBlockCosts::new(
+                Ok(DaBlockCosts {
+                    l2_block_range: 1..2,
+                    blob_cost_wei: 9000,
+                    blob_size_bytes: 3000,
+                }),
+                notifier.clone(),
+            ),
+            Some(Duration::from_millis(100)),
+        );
+        let mut watcher = StateWatcher::default();
+
+        let mut service = GasPriceServiceV1::new(
+            l2_block_source,
+            metadata_storage,
+            shared_algo,
+            algo_updater,
+            da_source,
+        );
+        let read_algo = service.next_block_algorithm();
+        let initial_price = read_algo.next_gas_price();
+
+        // the RunnableTask depends on the handle passed to it for the da block cost source to already be running,
+        // which is the responsibility of the UninitializedTask in the `into_task` method of the RunnableService
+        // here we mimic that behaviour by running the da block cost service.
+        service
+            .da_source_adapter_handle
+            .run(&mut watcher)
+            .await
+            .unwrap();
+
+        // when
+        service.run(&mut watcher).await.unwrap();
+        l2_block_sender.send(l2_block).await.unwrap();
```

Comment:
Is the bug you're seeing the reason we aren't submitting a new DA block here?

Reply:
not seeing the bug anymore :)

Reply:
I think this test should be "sending" a block to the DA record service, not the `l2_block_sender`.

Reply:
I still need to trigger `update_da_gas_price()`, so I've added it in the `update_da_record_data` fn of the `AlgorithmV1` impl in c516ea9b95